Testing: Improve interaction between dtypes and ops decorators

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48428 Migrate some torch.fft tests to use OpInfos
* #48427 Register OpInfos for torch.fft transforms
* **#48426 Testing: Improve interaction between dtypes and ops decorators**

Tests are run on the intersection of the dtypes requested and the types that are
supported by the operator (or are _not_ if `unsupported_dtypes_only` is used).

Differential Revision: [D25205835](https://our.internmc.facebook.com/intern/diff/D25205835)

peterbell10 committed:
Testing: Improve interaction between dtypes and ops decorators

Tests are run on the union of the dtypes requested and the types that are
supported by the operator (or are _not_ if `unsupported_dtypes_only` is used).

[ghstack-poisoned]

peterbell10 committed:
Update on "Testing: Improve interaction between dtypes and ops decorators"


Tests are run on the union of the dtypes requested and the types that are
supported by the operator (or are _not_ if `unsupported_dtypes_only` is used).

[ghstack-poisoned]

peterbell10 commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -283,7 +283,11 @@ def instantiated_test(self, name=name, test=test_fn, dtype=dtype, op=op):
+                if dtypes is not None:
+                    want_supported = not test.unsupported_dtypes_only
+                    dtypes = [t for t in dtypes
+                              if op.supports_dtype(t, cls.device_type) == want_supported]
```
This new behaviour is useful for #48428 because while the transforms do technically support integer inputs, they are just up-cast to float so it's not worth testing all integer types exhaustively.

mruberry commented:
I like the idea of this PR but want to discuss how we model this interaction.

A few ideas:
- it'd be nice if ops have a way of saying, "these are all the dtypes I support, but these are the dtypes you should (generally) tests"
- the ops decorator could default to the latter dtypes, but there should be a way, just like the dtype list can be inverted, to return all the supported dtypes
- we could model this as a property on an op that's a list of dtypes which is intersected with the supported dtypes (which are per device_type)

For example,

https://github.com/pytorch/pytorch/blob/e9efd8df1bce365d2dddbbc9e9266b7432aee5b2/test/test_ops.py#L41

probably wants to test ALL the dtypes, but a test like

https://github.com/pytorch/pytorch/blob/e9efd8df1bce365d2dddbbc9e9266b7432aee5b2/test/test_unary_ufuncs.py#L299

should probably just run a few common dtypes.

Note that we already have the `@dtypes` decorator which composes with the `@ops` decorator by OVERRIDING the `@ops` dtype interaction. So this proposed mechanism would be a way for ops, as opposed to tests, to suggest the dtypes to test.

What would you think of an approach like that, @peterbell10?

peterbell10 commented:
That seems like a good idea, but I still think it makes sense to have `dtypes` not just completely override the dtypes from the `OpInfo`. An example might be a test for NaN handling. Such a test only makes sense for `floating_and_complex_types()`, but you still only want to test the dtypes that the operator actually supports. i.e. something like this
```python
dtypes = cls._get_dtypes(test)
if test.all_dtypes:
    assert dtypes is None, "ops(all_dtypes=True) and dtypes decorator are incompatible"
    dtypes = op.supported_dtypes(cls.device_type)

    if test.unsupported_dtypes_only:
        dtypes = set(get_all_dtypes()).difference(dtypes)
else:
    dtypes = op.tested_dtypes(cls.device_type) if dtypes is None else set(dtypes)

    if test.unsupported_dtypes_only:
        dtypes = dtypes.difference(op.supported_dtypes(cls.device_type))
```

Also note that if you use `@dtypes(...)` with `@ops(op_db, unsupported_dtypes_only=True)` then it currently inverts the list of dtypes which is really unexpected. The new behaviour is to filter supported dtypes out of the dtype list. So for example if `torch.fft` starts supporting half precision complex values on CUDA, then `test_fft_half_errors` will only run for `bfloat16` on CUDA:
```python
@dtypes(torch.half, torch.bfloat16)
@ops(spectral_funcs, unsupported_dtypes_only=True)
def test_fft_half_errors(self, device, dtype, op):
    pass
```

mruberry commented:
I agree with you that the current behavior of `@dtypes` and `unsupported_dtypes_only=True` is bizarre.

I kind of like `@dtypes` overriding the dtypes. Maybe we can support the other variants through more arguments to the @ops decorator? For example, `unsupported_dtypes_only`, `all_dtypes`, and `filter_to`, where the last argument will intersect the list of dtypes that each op supports with the list of dtypes provided to `filter_to` (or similar name). It'd be nice if trying to compose any of these options with the @dtypes decorator threw an error, too (which shouldn't be too hard to do, either). 

Would that design address all the issues?

peterbell10 committed:
Update on "Testing: Improve interaction between dtypes and ops decorators"


Tests are run on the intersection of the dtypes requested and the types that are
supported by the operator (or are _not_ if `unsupported_dtypes_only` is used).

[ghstack-poisoned]

peterbell10 committed:
Update on "Testing: Improve interaction between dtypes and ops decorators"


Tests are run on the intersection of the dtypes requested and the types that are
supported by the operator (or are _not_ if `unsupported_dtypes_only` is used).

[ghstack-poisoned]

peterbell10 committed:
Update on "Testing: Improve interaction between dtypes and ops decorators"


Tests are run on the intersection of the dtypes requested and the types that are
supported by the operator (or are _not_ if `unsupported_dtypes_only` is used).

[ghstack-poisoned]

peterbell10 commented:
@mruberry I've implemented your suggestion with `filter_to` instead named `dtype_filter` to be clear that it's meant to be dtypes. 

Also, since `all_dtypes` and `unsupported_dtypes_only` are mutually exclusive, what do you think about using enum instead?
```python
class OpDTypes(Enum):
   basic = 0 # Test the basic set a.k.a. tested_dtypes (default)
   supported = 1  # Test all supported dtypes
   unsupported = 2  # Test only unsupported dtypes
```

So the decorator's usage would be:
```python
@ops(op_db, dtypes=OpDTypes.unsupported)
def test_unsupported_dtypes(self, device, dtype, op):
    pass

@ops(op_db, dtypes=OpDTypes.supported)
def test_supported_dtypes(self, device, dtype, op):
    pass
```

(Open to better names)

mruberry commented:
> @mruberry I've implemented your suggestion with `filter_to` instead named `dtype_filter` to be clear that it's meant to be dtypes.

I agree we can discover a better name. "dtype_filter" might suggest that those are the dtypes we're removing, however, since filters typically remove impurities. That is, a filter for X typically removes X. The behavior we want is to leave the dtypes in. "filter_to" might not be much better. Naming is hard.

Other name ideas: "only_these_dtypes", "allowed_dtypes"? 

> 
> Also, since `all_dtypes` and `unsupported_dtypes_only` are mutually exclusive, what do you think about using enum instead?
> 
> ```python
> class OpDTypes(Enum):
>    basic = 0 # Test the basic set a.k.a. tested_dtypes (default)
>    supported = 1  # Test all supported dtypes
>    unsupported = 2  # Test only unsupported dtypes
> ```
> 
> So the decorator's usage would be:
> 
> ```python
> @ops(op_db, dtypes=OpDTypes.unsupported)
> def test_unsupported_dtypes(self, device, dtype, op):
>     pass
> 
> @ops(op_db, dtypes=OpDTypes.supported)
> def test_supported_dtypes(self, device, dtype, op):
>     pass
> ```
> 
> (Open to better names)

That looks great. So if unsupported was selected and the unnamed option above were used, it would intersect the list of dtypes provided to the unnamed option with the unsupported dtypes?

peterbell10 commented:
> "dtype_filter" might suggest that those are the dtypes we're removing, however, since filters typically remove impurities.

How about `dtype_mask`, that or `allowed_dtypes` are my current favorites.

> if unsupported was selected and the unnamed option above were used, it would intersect the list of dtypes provided to the unnamed option with the unsupported dtypes?

Yeah that's what I think should happen. The enum determines the set of dtypes the operator wants to be tested on (in this case testing for lack of support), the mask/allow-list is the set of dtypes the test knows it can run. So, it's the intersection of these that should be tested.

mruberry commented:
> How about `dtype_mask`, that or `allowed_dtypes` are my current favorites.

"allowed_dtypes" sounds like a winner.
 
> > if unsupported was selected and the unnamed option above were used, it would intersect the list of dtypes provided to the unnamed option with the unsupported dtypes?
> 
> Yeah that's what I think should happen. The enum determines the set of dtypes the operator wants to be tested on (in this case testing for lack of support), the mask/allow-list is the set of dtypes the test knows it can run. So, it's the intersection of these that should be tested.

Sounds great.

peterbell10 committed:
Update on "Testing: Improve interaction between dtypes and ops decorators"


Tests are run on the intersection of the dtypes requested and the types that are
supported by the operator (or are _not_ if `unsupported_dtypes_only` is used).

[ghstack-poisoned]

peterbell10 commented:
Okay @mruberry, `dtype_filter` has been renamed `allowed_dtypes` and I've changed from flags to an enum.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -64,6 +64,7 @@ def __init__(self,
                  dtypesIfCPU=None,  # dtypes this function is expected to work with on CPU
                  dtypesIfCUDA=None,  # dtypes this function is expected to work with on CUDA
                  dtypesIfROCM=None,  # dtypes this function is expected to work with on ROCM
+                 tested_dtypes=None,  # dtypes to test with by default
```
How about "tested_dtypes" -> "default_test_dtypes"?

And the comment can elaborate that this iterable is intersected with the device-specific dtype iterables above.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -133,15 +135,22 @@ def should_skip(self, cls_name, test_name, device_type, dtype):
+        return dtype in self.supported_dtypes(device_type)
 
-        return dtype in self.dtypes
+    def tested_dtypes(self, device_type):
```
Name would be updated here, too. 

Add a comment here elaborating on what this function does, and then point people to the enum, which can have a more descriptive comment about the thinking behind each option, how they're intended to be used, and what the default behavior is.

mruberry commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -482,6 +483,13 @@ def instantiate_device_type_tests(generic_test_class, scope, except_for=None, on
 
 
+# Category of dtypes to run an OpInfo-based test for
+class OpDTypes(Enum):
```
Elaborate here on on how a test writer should choose between these options. We want to be able to direct developers less familiar with PyTorch's test suite to this comment and have them be able to choose the correct option.

mruberry commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -482,6 +483,13 @@ def instantiate_device_type_tests(generic_test_class, scope, except_for=None, on
 
+# Category of dtypes to run an OpInfo-based test for
+class OpDTypes(Enum):
+    basic = 0  # Test the basic set of dtypes (default)
```
"basic" -> "default" ?

mruberry commented on test/test_unary_ufuncs.py:
```diff
@@ -433,7 +433,7 @@ def compare_out(op, input, out):
                 op(input, out=output)
 
-    @ops(unary_ufuncs)
+    @ops(unary_ufuncs, dtypes=OpDTypes.supported)
```
Is this only a forward-looking change since no ops define their "default/basic" test dtypes yet?

mruberry commented:
PR looks good, @peterbell10! I have a few minor questions/suggestions. 

Why don't we test this behavior by having acos run most of its tests in long, half, bfloat16, float32, and cfloat? 

In the future I imagine those are the default dtypes we'll want for every unary ufunc. Although some unary ufuncs may want to override that list to include bool or uint8.

peterbell10 committed:
Update on "Testing: Improve interaction between dtypes and ops decorators"


Tests are run on the intersection of the dtypes requested and the types that are
supported by the operator (or are _not_ if `unsupported_dtypes_only` is used).

[ghstack-poisoned]

peterbell10 commented:
@mruberry PTAL. Renamed to `default_test_dtypes`, added more comments, and defined `default_test_dtypes` for `acos`.

mruberry reviewed (approved):
Cool; thanks @peterbell10!

peterbell10 committed:
Update on "Testing: Improve interaction between dtypes and ops decorators"


Tests are run on the intersection of the dtypes requested and the types that are
supported by the operator (or are _not_ if `unsupported_dtypes_only` is used).

Differential Revision: [D25205835](https://our.internmc.facebook.com/intern/diff/D25205835)

[ghstack-poisoned]

mruberry commented:
Hey @peterbell10, hate to bother you but would you mind rebasing this again? The internal tools are struggling with it for some reason.

peterbell10 committed:
Update on "Testing: Improve interaction between dtypes and ops decorators"


Tests are run on the intersection of the dtypes requested and the types that are
supported by the operator (or are _not_ if `unsupported_dtypes_only` is used).

Differential Revision: [D25205835](https://our.internmc.facebook.com/intern/diff/D25205835)

[ghstack-poisoned]

peterbell10 commented:
@mruberry I've rebased this.
